src/runtime/vm/bytecode.cc

```diff
@@ -252,13 +252,17 @@ Instruction::~Instruction() {
     case Opcode::GetTag:
     case Opcode::Goto:
     case Opcode::LoadConsti:
-    case Opcode::AllocStorage:
     case Opcode::ShapeOf:
     case Opcode::ReshapeTensor:
     case Opcode::DeviceCopy:
     case Opcode::Fatal:
     case Opcode::KillRegister:
       return;
+    case Opcode::AllocStorage:
+      if (this->alloc_storage.ndim > 0) {
```

Comment:
I'd prefer more direct

`this->alloc_storage.shape != nullptr`

since `this->alloc_storage.ndim > 0` is assuming the logic implemented elsewhere that the shape is non-null only when `this->alloc_storage.ndim > 0`.

Reply:
I'm afraid that check on `nullptr` might be not enough, because `shape` and `allocation_size` are in the same `union`. When we change the value of `allocation_size` then `shape` is not equal to `nullptr` and when we try to release "allocated" memory in `shape` then the program might be crashed with segfault.

With current implementation of `alloc_storage` instruction, we allocate memory for shape only in case when `ndim > 0` so this is why I think that it is a valid condition for releasing memory. 

If you have any ideas how to implement this logic more clearly, then let me know.